### PR TITLE
Add TWT ratio tap changer modification

### PR DESCRIPTION
**Pull request by ayolab**

### CODE REVIEW

File: src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java [hunk 1]
```diff
@@ -103,6 +118,249 @@ protected void modifyCharacteristics(Branch<?> branch, BranchModificationInfos b
...
+            tapChangerReports.add(Report.builder()
+                .withKey("tapsModification")
+                .withDefaultMessage("        Taps")
+                .withSeverity(TypedValue.INFO_SEVERITY)
+                .build());
+            addTapchangerSteps(tapChangerReports, tapChangerModificationInfos, adder);
+        } else if (tapChanger != null) {
+            for (Object step : tapChanger.getAllSteps().values()) {
+                if (step instanceof RatioTapChangerStep) {
+                    RatioTapChangerStep ratioStep = (RatioTapChangerStep) step;
```
Comment (dbraquart) [inline src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java:231]: this line can be deleted if you use:  
if (step instanceof RatioTapChangerStep ratioStep) {

File: src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java [hunk 2]
```diff
@@ -103,6 +118,249 @@ protected void modifyCharacteristics(Branch<?> branch, BranchModificationInfos b
...
+                    ((RatioTapChangerAdder) adder).beginStep().setR(ratioStep.getR()).setX(ratioStep.getX()).setG(ratioStep.getG()).setB(ratioStep.getB()).setRho(ratioStep.getRho()).endStep();
+                } else if (step instanceof PhaseTapChangerStep) {
+                    PhaseTapChangerStep phaseStep = (PhaseTapChangerStep) step;
```
Comment (dbraquart) [inline src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java:234]: same trick here

File: src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java [hunk 3]
```diff
@@ -103,6 +118,249 @@ protected void modifyCharacteristics(Branch<?> branch, BranchModificationInfos b
...
+                .withSeverity(TypedValue.INFO_SEVERITY)
+                .build());
+        for (TapChangerStepCreationInfos step : tapChangerModificationInfos.getSteps()) {
+            addStepAttributeReport(tapChangerStepsReports, "newStepIndex" + step.getIndex(),
+                    "            Tap (${index})", "index", String.valueOf(step.getIndex()));
+            addStepAttributeReport(tapChangerStepsReports, "newStepResistance" + step.getR(),
+                    "                Δ resistance : ${r}", "r", String.valueOf(step.getR()));
+            addStepAttributeReport(tapChangerStepsReports, "newStepReactance" + step.getX(),
+                                    "                Δ reactance : ${x}", "x", String.valueOf(step.getX()));
+            addStepAttributeReport(tapChangerStepsReports, "newStepConductance" + step.getX(),
```
Comment (dbraquart) [inline src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java:157]: must be step.getG() here

**Review by archendel — APPROVED** [review]

**Review by dbraquart — CHANGES_REQUESTED** [review]
code: simplify 2 cast
tests: cf Front PR

File: src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java [hunk 4]
```diff
@@ -103,6 +118,249 @@ protected void modifyCharacteristics(Branch<?> branch, BranchModificationInfos b
...
+            tapChangerReports.add(Report.builder()
+                .withKey("tapsModification")
+                .withDefaultMessage("        Taps")
+                .withSeverity(TypedValue.INFO_SEVERITY)
+                .build());
+            addTapchangerSteps(tapChangerReports, tapChangerModificationInfos, adder);
+        } else if (tapChanger != null) {
+            for (Object step : tapChanger.getAllSteps().values()) {
+                if (step instanceof RatioTapChangerStep) {
+                    RatioTapChangerStep ratioStep = (RatioTapChangerStep) step;
```
[the remark this answers is not in the document]
Comment (dbraquart) [reply src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java:231]: you can combine 2 lines in a single one

**Review by dbraquart — APPROVED** [review]
code ok
tests: done, see Front's PR

**Pull request merged**